src/core/common/data.cpp

```diff
@@ -0,0 +1,57 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements `Data` related function
+ */
+
+#include "data.hpp"
+
+namespace ot {
+
+bool CompareBytes(const uint8_t *aFirstBuffer, const uint8_t *aSecondBuffer, uint16_t aLength, ByteMatcher aMatcher)
+{
+    bool matches = true;
+
+    if (aMatcher == nullptr)
+    {
+        matches = (memcmp(aFirstBuffer, aSecondBuffer, aLength) == 0);
+        ExitNow();
+    }
+
+    while (aLength-- != 0)
```

Comment:
There is small advantage that it may lead to smaller code size and potentially faster looping. Tough most toolchain may optimize for this already.

The idea is that comparison against zero typically baked into assembly instructions (zero is often a fixed/known constant in most assembly instructions) and may be performed faster. Also in some toolchains, the zero comparison may come for free as part of the same assembly instruction which does the decrement (`--`).

Reply:
Thanks for explanation.